packages/medusa/src/services/fulfillment.ts

```diff
@@ -224,11 +234,34 @@ class FulfillmentService extends TransactionBaseService {
 
       const created = await Promise.all(
         fulfillments.map(async ({ shipping_method, items }) => {
+          await pvInventoryServiceTx.validateInventoryAtLocation(
+            items,
+            locationId!
+          )
+
+          await Promise.all(
+            items.map(async (i) => {
+              await pvInventoryServiceTx.adjustReservationsQuantityByLineItem(
+                i.id,
+                i.variant_id!,
+                locationId!,
+                -i.quantity
+              )
+
+              await pvInventoryServiceTx.adjustInventory(
+                i.variant_id!,
+                locationId!,
+                -i.quantity
+              )
+            })
+          )
+
```

Comment:
thought **(non-blocking)**: how do we handle something like this in a distributed fashion? (cc: @carlos-r-l-rodrigues)

E.g., imagine the following:
1. Order has Item A and Item B
2. Validation on line 237 succeeds
3. Immediately after there is a change in inventory for Item B (from an external system; e.g. an offline sale that updates the business's IMS system)
4. `adjustInventory` on line 251 now succeeds for Item A, but fails for Item B
5. Ideally, we would want to roll back the inventory adjustment for Item A, and throw an error.

If this wasn't handled internally in the FulfillmentService this would be easy to wrap in the TransactionOrchestrator, but not sure how we can manage this within services? 

---

Maybe I just answered my own question - should we consider moving the adjustment part out of the FulfillmentService and into the API handler instead?

Reply:
There are 2 ways we can solve that.

1st:
Handle it in batches. If one operation fails, everything will fail together and the local db transaction of the service will rollback everything.
This entire logic to `adjustReservationsQuantityByLineItem` and `adjustInventory`should be done in one call to the service.

2nd:
Using the transaction orchestrator, when a Promise.allSettled() finishes, if there is any rejected Promise, we revert all the fulfilled ones.